engine/routes.py

```diff
@@ -84,6 +86,114 @@ def decorated_function(*args, **kwargs):
     return inner
 
 
+def set_inactivity_timeout(is_active=lambda _: False, poll_period=None, max_runtime=None):
+    """
+    Decorator to poll whether a function is active and when it becomes inactive, to terminate it. Function activity is
+    checked by an is_active function passed to the decorator. Activity is checked at the end of every poll_period
+    seconds. Also, at the close of each poll period, if the cumulative runtime of the function exceeds max_runtime
+    seconds, the function is terminated. The is_active function has access to poll_period, max_runtime, and all
+    arguments passed to the wrapped function.
+    :param is_active: function used to determine if the decorated function is still active
+                      (e.g. if aggregation is progressing)
+    :param poll_period: the frequency, in seconds, at which the is_active function is called and the decorated
+                        function's runtime is checked against max_runtime
+    :param max_runtime: the maximum amount of time, in seconds, that the decorated function should be allowed to run;
+                        only checked at the end of each poll_period, so ideally max_runtime would be a factor of
+                        poll_period, otherwise the exact max_runtime will NOT be respected; a value of None indicates
+                        no max_runtime (i.e. never terminate function for running too long if it is still active)
+    :return: the object returned by the decorated function; WARNING: this will not work right for return values that
```

Comment:
What happens is the return value serialization fails?

Reply:
As written, if pickle.dumps fails, it should throw an exception which will be logged in stream engine. The child process will then be finished and cleaned up when the parent waits for it. Then the parent process will get an EOFError on pickle.loads, which will also be logged in stream engine. Finally, an "Unexpected internal error during request" message will go to EDEX. I can update the documentation to say an exception will be thrown in such cases.